src/iosrv/drpc_progress.c

```diff
@@ -272,6 +282,111 @@ destroy_session_node(struct drpc_list *session_node)
 	D_FREE(session_node);
 }
 
+/**
+ * ULT to execute the dRPC handler and send the response back
+ */
+static int
+drpc_handler_ult(void *call_ctx)
+{
+	int			rc;
+	struct drpc_call_ctx	*ctx = (struct drpc_call_ctx *)call_ctx;
+
+	D_INFO("dRPC handler ULT for module=%u method=%u\n",
+	       ctx->call->module, ctx->call->method);
+
+	ctx->session->handler(ctx->call, ctx->resp);
+
+	rc = drpc_send_response(ctx->session, ctx->resp);
+	if (rc != 0)
+		D_ERROR("Failed to send dRPC response (module=%u method=%u)\n",
+			ctx->call->module, ctx->call->method);
+
+	/*
+	 * We are responsible for cleaning up the call ctx.
+	 * Session is a copy - we can free it, but not close it. The main loop
+	 * still manages the sessions.
+	 */
+	drpc_free(ctx->session);
+	drpc_call_free(ctx->call);
+	drpc_response_free(ctx->resp);
+	D_FREE(ctx);
+
+	return rc;
+}
+
+static struct drpc_call_ctx *
+create_call_ctx(struct drpc *session_ctx, Drpc__Call *call,
+		Drpc__Response *resp)
+{
+	struct drpc_call_ctx *call_ctx;
+
+	D_ALLOC_PTR(call_ctx);
+	if (call_ctx == NULL)
+		return NULL;
+
+	/*
+	 * Need to copy the session context to avoid the main loop potentially
+	 * freeing that memory out from under our thread.
+	 */
+	call_ctx->session = drpc_dup(session_ctx);
+	if (call_ctx->session == NULL) {
+		D_FREE(call_ctx);
+		return NULL;
+	}
+
+	call_ctx->call = call;
+	call_ctx->resp = resp;
+
+	return call_ctx;
+}
+
+static int
+handle_incoming_call(struct drpc *session_ctx)
+{
+	int			rc;
+	Drpc__Call		*call = NULL;
+	Drpc__Response		*resp;
+	struct drpc_call_ctx	*call_ctx;
+
+	rc = drpc_recv_call(session_ctx, &call);
+	if (rc != 0)
+		return rc;
+
+	resp = drpc_response_create(call);
+	if (resp == NULL) {
+		D_ERROR("Could not allocate Drpc__Response\n");
+		drpc_call_free(call);
+		return -DER_NOMEM;
+	}
+
+	/*
+	 * Call and response become part of the call context - freeing will
+	 * be handled by the ULT.
+	 */
+	call_ctx = create_call_ctx(session_ctx, call, resp);
+	if (call_ctx == NULL) {
+		drpc_call_free(call);
+		drpc_response_free(resp);
+		return -DER_NOMEM;
+	}
+
+	/*
+	 * Ownership of the call context is passed on to the handler ULT.
+	 */
+	rc = dss_ult_create_execute(drpc_handler_ult, (void *)call_ctx,
+				    NULL, NULL,
```

Comment:
In this synchronous mode, I think this listener xstream will busy polling while waiting for drpc_handler_ult to complete. Any issues with dss_ult_create?

Reply:
sorry, I did not get the point why need to create the drpc_handler_ult for synchronous executing? As for synchronous, seems it is same as directly calls drpc_handler_ult func without creating ULT.

Reply:
We don't want synchronous execution... We want to spin off a ULT so the listener can go on listening while the command is executed in another thread. So I believe I did this wrong, and we should be using dss_ult_create as Li Wei suggested. Is that correct?

Reply:
in your current patch you pass NULL cb() for dss_ult_create_execute() so it is synchronous (see the code it internally wait its completion in that case).
yes dss_ult_create is async.

Reply:
Addressed this in the latest version of the patch.
I still need to add a routine to free ABT_threads that have finished. I was planning to have this be part of drpc_progress(), performed after dealing with the polling. However, in the current version of this patch, we block indefinitely - meaning the previous ABT_thread may not be freed until the next dRPC call comes in. Could be a long time, or never. I think we will need to have a finite timeout if we want the ABT artifacts cleaned up in a timely fashion.
Alternately we could spin off a ULT to periodically do the cleanup. That does introduce some complexity, worrying about concurrency on the drpc_progress_context, which is currently used in a single thread at a time. I don't know that it gains us anything, either, since we'll periodically have to eat some CPU cycles either way. @liw Thoughts?

Reply:
Latest version lets ABT clean up the threads on its own, rather than having us track them. Unless there's something else compelling for us to do with them, I'm inclined to keep it simple.